ui/prometheus-plugin/src/plugins/prometheus-time-series-query/PrometheusTimeSeriesQueryEditor.tsx

```diff
@@ -11,26 +11,20 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import { ChangeEvent } from 'react';
-import produce from 'immer';
+import { produce } from 'immer';
 import { Box, TextField, FormControl, InputLabel } from '@mui/material';
-import { OptionsEditorProps, DatasourceSelect, DatasourceSelectProps } from '@perses-dev/plugin-system';
-import { DEFAULT_PROM, isDefaultPromSelector, isPrometheusDatasourceSelector } from '../model';
-import { PrometheusTimeSeriesQuerySpec } from './time-series-query';
-
-export type PrometheusTimeSeriesQueryEditorProps = OptionsEditorProps<PrometheusTimeSeriesQuerySpec>;
+import { DatasourceSelect, DatasourceSelectProps } from '@perses-dev/plugin-system';
+import { DEFAULT_PROM, isDefaultPromSelector, isPrometheusDatasourceSelector } from '../../model';
+import { PrometheusTimeSeriesQueryEditorProps, useQueryState } from './query-editor-model';
 
+/**
+ * The options editor component for editing a PrometheusTimeSeriesQuery's spec.
+ */
 export function PrometheusTimeSeriesQueryEditor(props: PrometheusTimeSeriesQueryEditorProps) {
   const { onChange, value } = props;
-  const { query, datasource } = value;
+  const { datasource } = value;
 
-  const handleQueryChange = (e: ChangeEvent<HTMLInputElement>) => {
-    onChange(
-      produce(value, (draft) => {
-        draft.query = e.target.value;
-      })
-    );
-  };
+  const { query, handleQueryChange, handleQueryBlur } = useQueryState(props);
```

Comment:
Just realized, we could get verbose and call it `usePrometheusTimeSeriesQueryState(props)`, but I'll leave that deicison up to you (to verbose or not to verbose... to be or not to be)

Reply:
😂  If the hook was getting exported for use elsewhere, I'd probably go verbose. But since it's just for internal use by this one component, I think the shorter version works all right.